File: crates/harness-tools/src/fs.rs
//! Sandboxed filesystem tools. Every tool is constructed with a `root`
//! directory and refuses inputs that are absolute or contain `..`
//! components, so the LLM can only read/list/write inside that root.

use std::path::PathBuf;

use async_trait::async_trait;
use harness_core::{BoxError, Tool};
use serde_json::{json, Value};
use tokio::fs;

use crate::sandbox::resolve_under;

fn arg_str<'a>(args: &'a Value, key: &str) -> Result<&'a str, BoxError> {
    args.get(key)
        .and_then(Value::as_str)
        .ok_or_else(|| -> BoxError { format!("missing `{key}` argument").into() })
}

/// Read a UTF-8 file under the tool root.
pub struct FsReadTool {
    root: PathBuf,
}

impl FsReadTool {
    pub fn new(root: impl Into<PathBuf>) -> Self {
        Self { root: root.into() }
    }
}

#[async_trait]
impl Tool for FsReadTool {
    fn name(&self) -> &str {
        "fs.read"
    }

    fn description(&self) -> &str {
        "Read a UTF-8 text file located under the tool's root directory. \
         `path` is relative; `..` and absolute paths are rejected."
    }

    fn parameters(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "path": { "type": "string", "description": "Relative path under the root." }
            },
            "required": ["path"]
        })
    }

    fn cacheable(&self) -> bool {
        true
    }

    async fn invoke(&self, args: Value) -> Result<String, BoxError> {
        let rel = arg_str(&args, "path")?;
        let abs = resolve_under(&self.root, rel)?;
        let contents = fs::read_to_string(&abs).await?;
        Ok(contents)
    }
}

/// List entries in a directory under the tool root.
pub struct FsListTool {
    root: PathBuf,
}

impl FsListTool {
    pub fn new(root: impl Into<PathBuf>) -> Self {
        Self { root: root.into() }
    }
}

#[async_trait]
impl Tool for FsListTool {
    fn name(&self) -> &str {
        "fs.list"
    }

    fn description(&self) -> &str {
        "List entries in a directory under the tool root. Returns a JSON array \
         of {name, kind: 'file'|'dir'|'other'} objects."
    }

    fn parameters(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": "Relative directory path under the root. \
                                    Defaults to the root itself."
                }
            }
        })
    }

    fn cacheable(&self) -> bool {
        true
    }

    async fn invoke(&self, args: Value) -> Result<String, BoxError> {
        let rel = args.get("path").and_then(Value::as_str).unwrap_or(".");
        let abs = resolve_under(&self.root, rel)?;
        let mut rd = fs::read_dir(&abs).await?;

        let mut entries = Vec::new();
        while let Some(entry) = rd.next_entry().await? {
            let name = entry.file_name().to_string_lossy().into_owned();
            let ft = entry.file_type().await?;
            let kind = if ft.is_file() {
                "file"
            } else if ft.is_dir() {
                "dir"
            } else {
                "other"
            };
            entries.push(json!({ "name": name, "kind": kind }));
        }
        Ok(Value::Array(entries).to_string())
    }
}

/// Write UTF-8 `content` to `path`, creating parent directories as needed.
///
/// Overwrites existing files. Intentionally not registered by default in
/// `register_builtins` — opt in via `BuiltinsConfig::enable_fs_write`.
pub struct FsWriteTool {
    root: PathBuf,
}

impl FsWriteTool {
    pub fn new(root: impl Into<PathBuf>) -> Self {
        Self { root: root.into() }
    }
}

#[async_trait]
impl Tool for FsWriteTool {
    fn name(&self) -> &str {
        "fs.write"
    }

    fn requires_approval(&self) -> bool {
        true
    }

    fn description(&self) -> &str {
        "Write UTF-8 text to a file under the tool root. Creates parent \
         directories as needed and overwrites existing files."
    }

    fn parameters(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "path": { "type": "string", "description": "Relative path under the root." },
                "content": { "type": "string", "description": "File contents (UTF-8)." }
            },
            "required": ["path", "content"]
        })
    }

    fn cacheable(&self) -> bool {
        true
    }

    async fn invoke(&self, args: Value) -> Result<String, BoxError> {
        let rel = arg_str(&args, "path")?;
        let content = arg_str(&args, "content")?;
        let abs = resolve_under(&self.root, rel)?;
        if let Some(parent) = abs.parent() {
            fs::create_dir_all(parent).await?;
        }
        let bytes = content.len();
        fs::write(&abs, content).await?;
        Ok(format!("wrote {bytes} bytes to {}", abs.display()))
    }
}

/// Replace `old_string` with `new_string` in a UTF-8 file under the tool
/// root. By default the match must be unique so the LLM can't silently
/// rewrite many call-sites at once; pass `replace_all = true` to opt in.
///
/// Intentionally not registered by default in `register_builtins` — opt in
/// via `BuiltinsConfig::enable_fs_edit`.
pub struct FsEditTool {
    root: PathBuf,
}

impl FsEditTool {
    pub fn new(root: impl Into<PathBuf>) -> Self {
        Self { root: root.into() }
    }
}

#[async_trait]
impl Tool for FsEditTool {
    fn name(&self) -> &str {
        "fs.edit"
    }

    fn requires_approval(&self) -> bool {
        true
    }

    fn description(&self) -> &str {
        "Replace `old_string` with `new_string` in a file under the tool root. \
         By default `old_string` must occur exactly once; pass \
         `replace_all = true` to replace every occurrence."
    }

    fn parameters(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "path": { "type": "string", "description": "Relative path under the root." },
                "old_string": { "type": "string", "description": "Exact text to replace." },
                "new_string": { "type": "string", "description": "Replacement text." },
                "replace_all": {
                    "type": "boolean",
                    "description": "Replace every occurrence instead of requiring uniqueness. Defaults to false."
                }
            },
            "required": ["path", "old_string", "new_string"]
        })
    }

    fn cacheable(&self) -> bool {
        true
    }

    async fn invoke(&self, args: Value) -> Result<String, BoxError> {
        let rel = arg_str(&args, "path")?;
        let old = arg_str(&args, "old_string")?;
        let new = arg_str(&args, "new_string")?;
        let replace_all = args
            .get("replace_all")
            .and_then(Value::as_bool)
            .unwrap_or(false);

        if old.is_empty() {
            return Err("`old_string` must not be empty".into());
        }
        if old == new {
            return Err("`old_string` and `new_string` are identical".into());
        }

        let abs = resolve_under(&self.root, rel)?;
        let original = fs::read_to_string(&abs).await?;
        let count = original.matches(old).count();

        let updated = match (count, replace_all) {
            (0, _) => return Err("`old_string` not found in file".into()),
            (_, true) => original.replace(old, new),
            (1, false) => original.replacen(old, new, 1),
            (n, false) => {
                return Err(format!(
                    "`old_string` matches {n} times — pass `replace_all = true` or extend the snippet for uniqueness"
                )
                .into());
            }
        };

        fs::write(&abs, &updated).await?;
        let replaced = if replace_all { count } else { 1 };
        Ok(format!(
            "edited {}: replaced {replaced} occurrence(s)",
            abs.display()
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[tokio::test]
    async fn read_write_roundtrip() {
        let dir = tempdir().unwrap();
        let write = FsWriteTool::new(dir.path());
        let read = FsReadTool::new(dir.path());

        write
            .invoke(json!({ "path": "sub/hello.txt", "content": "world" }))
            .await
            .unwrap();
        let got = read.invoke(json!({ "path": "sub/hello.txt" })).await.unwrap();
        assert_eq!(got, "world");
    }

    #[tokio::test]
    async fn list_contains_written_file() {
        let dir = tempdir().unwrap();
        let write = FsWriteTool::new(dir.path());
        let list = FsListTool::new(dir.path());
        write
            .invoke(json!({ "path": "a.txt", "content": "x" }))
            .await
            .unwrap();
        let out = list.invoke(json!({})).await.unwrap();
        assert!(out.contains("a.txt"), "list output was {out}");
    }

    #[tokio::test]
    async fn edit_replaces_unique_match() {
        let dir = tempdir().unwrap();
        let write = FsWriteTool::new(dir.path());
        let edit = FsEditTool::new(dir.path());
        let read = FsReadTool::new(dir.path());

        write
            .invoke(json!({ "path": "f.txt", "content": "hello world" }))
            .await
            .unwrap();
        edit.invoke(json!({
            "path": "f.txt",
            "old_string": "world",
            "new_string": "rust"
        }))
        .await
        .unwrap();
        let got = read.invoke(json!({ "path": "f.txt" })).await.unwrap();
        assert_eq!(got, "hello rust");
    }

    #[tokio::test]
    async fn edit_rejects_ambiguous_match() {
        let dir = tempdir().unwrap();
        let write = FsWriteTool::new(dir.path());
        let edit = FsEditTool::new(dir.path());

        write
            .invoke(json!({ "path": "f.txt", "content": "ab ab ab" }))
            .await
            .unwrap();
        let err = edit
            .invoke(json!({
                "path": "f.txt",
                "old_string": "ab",
                "new_string": "x"
            }))
            .await
            .unwrap_err();
        assert!(err.to_string().contains("matches 3"), "got: {err}");
    }

    #[tokio::test]
    async fn edit_replace_all() {
        let dir = tempdir().unwrap();
        let write = FsWriteTool::new(dir.path());
        let edit = FsEditTool::new(dir.path());
        let read = FsReadTool::new(dir.path());

        write
            .invoke(json!({ "path": "f.txt", "content": "ab ab ab" }))
            .await
            .unwrap();
        edit.invoke(json!({
            "path": "f.txt",
            "old_string": "ab",
            "new_string": "x",
            "replace_all": true
        }))
        .await
        .unwrap();
        let got = read.invoke(json!({ "path": "f.txt" })).await.unwrap();
        assert_eq!(got, "x x x");
    }

    #[tokio::test]
    async fn edit_missing_string_errors() {
        let dir = tempdir().unwrap();
        let write = FsWriteTool::new(dir.path());
        let edit = FsEditTool::new(dir.path());

        write
            .invoke(json!({ "path": "f.txt", "content": "abc" }))
            .await
            .unwrap();
        let err = edit
            .invoke(json!({
                "path": "f.txt",
                "old_string": "xyz",
                "new_string": "q"
            }))
            .await
            .unwrap_err();
        assert!(err.to_string().contains("not found"), "got: {err}");
    }

    #[tokio::test]
    async fn edit_rejects_empty_old() {
        let dir = tempdir().unwrap();
        let write = FsWriteTool::new(dir.path());
        let edit = FsEditTool::new(dir.path());

        write
            .invoke(json!({ "path": "f.txt", "content": "abc" }))
            .await
            .unwrap();
        let err = edit
            .invoke(json!({
                "path": "f.txt",
                "old_string": "",
                "new_string": "q"
            }))
            .await
            .unwrap_err();
        assert!(err.to_string().contains("must not be empty"), "got: {err}");
    }
}

File: packages/tools/src/fs.ts
// Sandboxed filesystem tools. Every tool is constructed with a `root`
// directory and refuses inputs that are absolute or contain `..` components,
// so the LLM can only read/list/write inside that root.
//
// Ported faithfully from crates/harness-tools/src/fs.rs.

import { Buffer } from "node:buffer";
import * as fs from "node:fs/promises";
import * as path from "node:path";
import type { JsonValue, Tool, ToolCategory } from "@jarvis/core";
import { resolveUnder } from "./sandbox.ts";
import { withDiagnostics, type DiagnosticsHook } from "./diagnostics.ts";

/** Configuration shared by the `fs.*` tools. */
export interface FsConfig {
  /** Sandbox root; every path argument is resolved under it. */
  root: string;
  /**
   * Optional post-write diagnostics hook. The write tools (`fs.write`,
   * `fs.edit`) call it with the path they wrote and append the returned
   * `<diagnostics>` block. Read-only tools ignore it. Absent → no change.
   */
  diagnostics?: DiagnosticsHook;
}

function asObject(args: JsonValue): { [key: string]: JsonValue } {
  if (args !== null && typeof args === "object" && !Array.isArray(args)) {
    return args;
  }
  return {};
}

function argStr(args: JsonValue, key: string): string {
  const v = asObject(args)[key];
  if (typeof v !== "string") {
    throw new Error(`missing \`${key}\` argument`);
  }
  return v;
}

/**
 * Truncate a UTF-8 string to at most `max` bytes without splitting a
 * multi-byte sequence, mirroring the Rust char-boundary backtrack.
 */
function truncateBytes(contents: string, max: number): string {
  const buf = Buffer.from(contents, "utf8");
  if (buf.length <= max) return contents;
  let idx = max;
  // Back up off a UTF-8 continuation byte (0b10xxxxxx) so we cut on a char
  // boundary, matching `is_char_boundary` in the Rust.
  while (idx > 0 && (buf[idx]! & 0xc0) === 0x80) {
    idx -= 1;
  }
  const head = buf.subarray(0, idx).toString("utf8");
  return `${head}\n\n[... truncated at ${max} bytes ...]`;
}

/** Read a UTF-8 file under the tool root. */
export class FsReadTool implements Tool {
  readonly name = "fs.read";
  readonly description =
    "Read a UTF-8 text file located under the tool's root directory. " +
    "`path` is relative; `..` and absolute paths are rejected. " +
    "Files larger than the configured byte limit are truncated.";
  readonly cacheable = true;
  readonly category: ToolCategory = "read";

  #root: string;
  #maxBytes: number | undefined;

  constructor(config: FsConfig & { maxBytes?: number }) {
    this.#root = config.root;
    this.#maxBytes = config.maxBytes;
  }

  get parameters(): JsonValue {
    return {
      type: "object",
      properties: {
        path: { type: "string", description: "Relative path under the root." },
      },
      required: ["path"],
    };
  }

  async invoke(args: JsonValue): Promise<string> {
    const rel = argStr(args, "path");
    const abs = resolveUnder(this.#root, rel);
    const contents = await fs.readFile(abs, "utf8");
    if (this.#maxBytes !== undefined) {
      return truncateBytes(contents, this.#maxBytes);
    }
    return contents;
  }
}

/** List entries in a directory under the tool root. */
export class FsListTool implements Tool {
  readonly name = "fs.list";
  readonly description =
    "List entries in a directory under the tool root. Returns a JSON array " +
    "of {name, kind: 'file'|'dir'|'other'} objects.";
  readonly cacheable = true;
  readonly category: ToolCategory = "read";

  #root: string;

  constructor(config: FsConfig) {
    this.#root = config.root;
  }

  get parameters(): JsonValue {
    return {
      type: "object",
      properties: {
        path: {
          type: "string",
          description:
            "Relative directory path under the root. Defaults to the root itself.",
        },
      },
    };
  }

  async invoke(args: JsonValue): Promise<string> {
    const relRaw = asObject(args).path;
    const rel = typeof relRaw === "string" ? relRaw : ".";
    const abs = resolveUnder(this.#root, rel);
    const dirents = await fs.readdir(abs, { withFileTypes: true });

    const entries: JsonValue[] = [];
    for (const entry of dirents) {
      const kind = entry.isFile() ? "file" : entry.isDirectory() ? "dir" : "other";
      entries.push({ name: entry.name, kind });
    }
    return JSON.stringify(entries);
  }
}

/**
 * Write UTF-8 `content` to `path`, creating parent directories as needed.
 *
 * Overwrites existing files. Opt-in + approval-gated.
 */
export class FsWriteTool implements Tool {
  readonly name = "fs.write";
  readonly description =
    "Write UTF-8 text to a file under the tool root. Creates parent " +
    "directories as needed and overwrites existing files.";
  readonly requiresApproval = true;
  readonly cacheable = true;
  readonly category: ToolCategory = "write";

  #root: string;
  #diagnostics: DiagnosticsHook | undefined;

  constructor(config: FsConfig) {
    this.#root = config.root;
    this.#diagnostics = config.diagnostics;
  }

  get parameters(): JsonValue {
    return {
      type: "object",
      properties: {
        path: { type: "string", description: "Relative path under the root." },
        content: { type: "string", description: "File contents (UTF-8)." },
      },
      required: ["path", "content"],
    };
  }

  async invoke(args: JsonValue): Promise<string> {
    const rel = argStr(args, "path");
    const content = argStr(args, "content");
    const abs = resolveUnder(this.#root, rel);
    const parent = path.dirname(abs);
    await fs.mkdir(parent, { recursive: true });
    const bytes = Buffer.byteLength(content, "utf8");
    await fs.writeFile(abs, content, "utf8");
    return withDiagnostics(`wrote ${bytes} bytes to ${abs}`, this.#diagnostics, [abs]);
  }
}

/**
 * Replace `old_string` with `new_string` in a UTF-8 file under the tool root.
 * By default the match must be unique so the LLM can't silently rewrite many
 * call-sites at once; pass `replace_all = true` to opt in.
 *
 * Opt-in + approval-gated.
 */
export class FsEditTool implements Tool {
  readonly name = "fs.edit";
  readonly description =
    "Replace `old_string` with `new_string` in a file under the tool root. " +
    "By default `old_string` must occur exactly once; pass " +
    "`replace_all = true` to replace every occurrence.";
  readonly requiresApproval = true;
  readonly cacheable = true;
  readonly category: ToolCategory = "write";

  #root: string;
  #diagnostics: DiagnosticsHook | undefined;

  constructor(config: FsConfig) {
    this.#root = config.root;
    this.#diagnostics = config.diagnostics;
  }

  get parameters(): JsonValue {
    return {
      type: "object",
      properties: {
        path: { type: "string", description: "Relative path under the root." },
        old_string: { type: "string", description: "Exact text to replace." },
        new_string: { type: "string", description: "Replacement text." },
        replace_all: {
          type: "boolean",
          description:
            "Replace every occurrence instead of requiring uniqueness. Defaults to false.",
        },
      },
      required: ["path", "old_string", "new_string"],
    };
  }

  async invoke(args: JsonValue): Promise<string> {
    const rel = argStr(args, "path");
    const oldString = argStr(args, "old_string");
    const newString = argStr(args, "new_string");
    const replaceAllRaw = asObject(args).replace_all;
    const replaceAll = typeof replaceAllRaw === "boolean" ? replaceAllRaw : false;

    if (oldString.length === 0) {
      throw new Error("`old_string` must not be empty");
    }
    if (oldString === newString) {
      throw new Error("`old_string` and `new_string` are identical");
    }

    const abs = resolveUnder(this.#root, rel);
    const original = await fs.readFile(abs, "utf8");
    const count = countOccurrences(original, oldString);

    let updated: string;
    let replaced: number;
    if (count === 0) {
      throw new Error("`old_string` not found in file");
    } else if (replaceAll) {
      // split/join does a literal replacement (no `$`-pattern interpolation,
      // unlike String.prototype.replace with a string pattern).
      updated = original.split(oldString).join(newString);
      replaced = count;
    } else if (count === 1) {
      const idx = original.indexOf(oldString);
      updated = original.slice(0, idx) + newString + original.slice(idx + oldString.length);
      replaced = 1;
    } else {
      throw new Error(
        `\`old_string\` matches ${count} times — pass \`replace_all = true\` or extend the snippet for uniqueness`,
      );
    }

    await fs.writeFile(abs, updated, "utf8");
    return withDiagnostics(`edited ${abs}: replaced ${replaced} occurrence(s)`, this.#diagnostics, [abs]);
  }
}

/**
 * Count non-overlapping occurrences of `needle` in `haystack`, matching
 * Rust's `str::matches(...).count()`. `needle` is guaranteed non-empty by the
 * caller, but guard anyway.
 */
function countOccurrences(haystack: string, needle: string): number {
  if (needle.length === 0) return 0;
  let count = 0;
  let from = 0;
  for (;;) {
    const idx = haystack.indexOf(needle, from);
    if (idx === -1) break;
    count += 1;
    from = idx + needle.length;
  }
  return count;
}
